### PR TITLE
Support for value_and_grad at the MLIR side (inside a jitting context)

### PR DESCRIPTION
**Pull request by rauletorresc**
**Context:** Lowering of value_and_grad to MLIR is now supported when @qjit is used.

**Description of the Change:** A new ValueAndGradOp was added and the existing BackpropOp was modified to return two results: values and gradients. 

**Benefits:** Code reutilization.

**TODO:**
- [ ] Add unit tests.
- [ ] Test corner cases.
- [ ] Update the changelog.md file

[sc-55116]

### CODE REVIEW

File: frontend/catalyst/jax_primitives.py [hunk 1]
```diff
@@ -413,6 +415,63 @@ def _grad_lowering(ctx, *args, jaxpr, fn, grad_params):
...
+@value_and_grad_p.def_impl
+def _value_and_grad_def_impl(ctx, *args, jaxpr, fn, grad_params):  # pragma: no cover
+    raise NotImplementedError()
+
+
+@value_and_grad_p.def_abstract_eval
+def _value_and_grad_abstract(*args, jaxpr, fn, grad_params):  # pylint: disable=unused-argument
+    """This function is called with abstract arguments for tracing.
+    Note: argument names must match these of `_value_and_grad_lowering`."""
+    return jaxpr.out_avals + jaxpr.out_avals
```
Comment (erick-xanadu) [inline frontend/catalyst/jax_primitives.py:429]: Just for my curiousity (I am not too familiar with value and grad). Here JAXPR represents the function that will be derived and the out_avals are its outputs. Is the output always going to be the same types repeated? I remember when I implemented a little bit of gradients we had to adjust the signature based on which partial derivative we were taking. Maybe this is already taken into account and I just forgot.

File: mlir/lib/Gradient/Transforms/ConversionPatterns.cpp [hunk 2]
```diff
@@ -362,9 +362,15 @@ struct BackpropOpPattern : public ConvertOpToLLVMPattern<BackpropOp> {
             }
         }
 
+        // If we don't have any vals, it means we are coming from a GradOp, then we mark the results
+        // for being ignored by Enzime. But if we have vals, we are coming from a ValueAndGradOp,
+        // then we tell Enzime to keep them.
+        bool dupNoNeed = op.getVals().empty();
```
Comment (erick-xanadu) [inline mlir/lib/Gradient/Transforms/ConversionPatterns.cpp:368]: Nice!

File: frontend/catalyst/jax_primitives.py (hunk 1 above)
Comment (dime10) [reply frontend/catalyst/jax_primitives.py:429]: Yes great point, have a look at the abstract eval of grad, the outvals are correct here but the gradient types need to be deduced correctly.

**Comment by paul0403** [comment]
Closing this as `value_and_grad` is merged in by #804
